vyper/venom/passes/algebraic_optimization.py

```diff
@@ -57,24 +85,302 @@ def _get_iszero_chain(self, op: IROperand) -> list[IRInstruction]:
+        var = inst.parent.parent.get_next_variable()
+        operands = [arg if isinstance(arg, IROperand) else IRLiteral(arg) for arg in args]
+        new_inst = IRInstruction(opcode, operands, output=var)
+        inst.parent.insert_instruction(new_inst, index)
+        for op in new_inst.operands:
+            if isinstance(op, IRVariable):
+                self.dfg.add_use(op, new_inst)
+        self.dfg.add_use(var, inst)
+        self.dfg.set_producing_instruction(var, new_inst)
+        return var
+
+    def _is_lit(self, operand: IROperand) -> bool:
+        return isinstance(operand, IRLiteral)
+
+    def _lit_eq(self, operand: IROperand, val: int) -> bool:
+        return self._is_lit(operand) and operand.value == val
+
+    def _op_eq(self, operands, idx_a: int, idx_b: int) -> bool:
+        if self._is_lit(operands[idx_a]) and self._is_lit(operands[idx_b]):
+            return operands[idx_a].value == operands[idx_b].value
+        else:
+            return operands[idx_a] == operands[idx_b]
+
+    def _algebraic_opt(self):
+        self.last = False
+        while self._algebraic_opt_pass():
+            pass
+        self.last = True
+        self._algebraic_opt_pass()
+
+    def _algebraic_opt_pass(self) -> bool:
+        change = False
         for bb in self.function.get_basic_blocks():
             for inst in bb.instructions:
-                # check if the instruction is of the form
-                # `add <ptr> <label>`
-                # this works only if store chains have been eliminated.
-                if (
-                    inst.opcode == "add"
-                    and isinstance(inst.operands[0], IRLiteral)
-                    and isinstance(inst.operands[1], IRLabel)
-                ):
-                    inst.opcode = "offset"
+                if self._handle_inst_peephole(inst):
+                    change |= True
+
+        return change
+
+    def _handle_inst_peephole(self, inst: IRInstruction) -> bool:
+        if inst.opcode == "assert":
+            return self.handle_assert_inst(inst)
+        if inst.output is None:
+            return False
+        if inst.is_volatile:
+            return False
+        if inst.opcode == "store":
+            return False
+        if inst.is_pseudo:
+            return False
+
+        operands = inst.operands
+
+        if (
+            inst.opcode == "add"
+            and self._is_lit(operands[0])
+            and isinstance(inst.operands[1], IRLabel)
+        ):
+            inst.opcode = "offset"
+            return True
+
+        if inst.is_commutative and self._is_lit(operands[1]):
+            operands = [operands[1], operands[0]]
+
+        if inst.opcode in {"shl", "shr", "sar"}:
+            if self._lit_eq(operands[1], 0):
+                return self._store(inst, operands[0])
+            # no more cases for these instructions
+            return False
+
+        if inst.opcode in {"add", "sub", "xor"}:
+            if self._lit_eq(operands[0], 0):
```

Comment:
can you add comments stating the algebraic rule for each rule? e.g.
```suggestion
            # x + 0 == x - 0 == x ^ 0 == x
            if self._lit_eq(operands[0], 0):
```